daemon/config/config.go

```diff
@@ -110,3 +115,96 @@ func (cfg *Config) Validate() error {
 
 	return nil
 }
+
+//MergeConfigConfigurations merges flagSet flags and config file flags into Config.
+func (cfg *Config) MergeConfigConfigurations(config *Config, flagSet *pflag.FlagSet) error {
+	contents, err := ioutil.ReadFile(cfg.ConfigFile)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to read contents from config file %s: %s", cfg.ConfigFile, err)
+	}
+
+	var origin map[string]interface{}
+	if err = json.NewDecoder(bytes.NewReader(contents)).Decode(&origin); err != nil {
+		return fmt.Errorf("failed to decode json: %s", err)
+	}
+
+	fileFlags := make(map[string]interface{}, 0)
+	iterateConfig(origin, fileFlags)
+
+	if len(fileFlags) == 0 {
+		return nil
+	}
+
+	// check if invalid or unknown flag exist in config file
+	if err = getUnknownFlags(flagSet, fileFlags); err != nil {
+		return err
+	}
+
+	// check conflict in command line flags and config file
+	if err = getConflictConfigurations(flagSet, fileFlags); err != nil {
+		return err
+	}
+
+	fileConfig := &Config{}
+	if err = json.NewDecoder(bytes.NewReader(contents)).Decode(fileConfig); err != nil {
+		return fmt.Errorf("failed to decode json: %s", err)
+	}
+
+	// merge configurations from command line flags and config file
+	err = mergeConfigurations(fileConfig, cfg)
+	return err
+
+}
+
+// iterateConfig resolves key-value from config file iteratly.
+func iterateConfig(origin map[string]interface{}, config map[string]interface{}) {
```

Comment:
I think we need to make this function some kind of robust. If the config is nil, I am afraid this function would panic. How about doing this nil-checking to improve robustness?

Reply:
I refer to make this function before `iterateConfig`, so that origin can not be nil, and since it not a common function, will not be used by other function ,add nil check is redundancy

Reply:
I still insist on make functions independent from each other. As if you encapsulate this function out of the another place which is calling this. It has a trend to be general. If that, we cannot prevent others to use it in the future, as a result, there will be potential possibility to panic.

In addition, if the code which is calling this function changes or removes the nil checking in the upper layer, there will be potential panic as well.

Reply:
for range a nil map will never cause panic, since nil value will convert to a nil map(map[] ), that add a nil check is useless.